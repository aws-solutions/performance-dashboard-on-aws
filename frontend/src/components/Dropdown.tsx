/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";

interface Props {
    id: string;
    name: string;
    label: string;
    options: Array<{
        value: string;
        label: string;
    }>;
    defaultValue?: string;
    value?: string;
    register?: Function;
    hint?: string;
    error?: string;
    required?: boolean;
    onChange?: Function;
    className?: string;
}

function Dropdown(props: Props) {
    let formGroupClassName = "usa-form-group";
    if (props.error) {
        formGroupClassName += " usa-form-group--error";
    }
    if (props.className) {
        formGroupClassName += ` ${props.className}`;
    }

    const handleChange = (event: React.FormEvent<HTMLSelectElement>) => {
        if (props.onChange) {
            props.onChange(event);
        }
    };

    return (
        <div className={formGroupClassName} role="contentinfo">
            <label htmlFor={props.id} className="usa-label text-bold">
                {props.label}
            </label>
            <div id={`${props.id}-description`} className="usa-hint">
                {props.hint}
            </div>
            {props.error && (
                <span className="usa-error-message" id="input-error-message" role="alert">
                    {props.error}
                </span>
            )}
            <select
                id={props.id}
                aria-describedby={`${props.id}-description`}
                defaultValue={props.defaultValue}
                value={props.value}
                ref={props.register && props.register({ required: props.required })}
                name={props.name}
                onChange={handleChange}
                className="usa-select"
            >
                {props.options.map((option) => {
                    return (
                        <option key={option.value} value={option.value}>
                            {option.label}
                        </option>
                    );
                })}
            </select>
        </div>
    );
}

export default Dropdown;
