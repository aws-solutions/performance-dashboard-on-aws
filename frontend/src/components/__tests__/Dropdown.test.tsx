/*
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *  SPDX-License-Identifier: Apache-2.0
 */

import React from "react";
import { render } from "@testing-library/react";
import Dropdown from "../Dropdown";

test("renders an empty dropdown", async () => {
    const wrapper = render(
        <Dropdown id="aws-service" name="aws-service" label="Select an AWS service" options={[]} />,
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("renders a dropdown with options", async () => {
    const wrapper = render(
        <Dropdown
            id="aws-service"
            name="aws-service"
            label="Select an AWS service"
            options={[
                {
                    label: "Amazon S3",
                    value: "s3",
                },
                {
                    label: "AWS Lambda",
                    value: "lambda",
                },
            ]}
        />,
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("calls register function", async () => {
    const register = jest.fn();
    render(
        <Dropdown
            id="aws-service"
            name="aws-service"
            register={register}
            label="Select an AWS service"
            options={[]}
        />,
    );
    expect(register).toBeCalled();
});

test("renders a select with error", async () => {
    const wrapper = render(
        <Dropdown
            id="aws-service"
            name="aws-service"
            error="Something bad happened"
            label="Select an AWS service"
            options={[]}
        />,
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("renders a select with hint", async () => {
    const wrapper = render(
        <Dropdown
            id="aws-service"
            name="aws-service"
            hint="Think twice before selecting"
            label="Select an AWS service"
            options={[]}
        />,
    );
    expect(wrapper.container).toMatchSnapshot();
});

test("renders a select with a default value", async () => {
    const wrapper = render(
        <Dropdown
            id="aws-service"
            name="aws-service"
            defaultValue="lambda"
            label="Select an AWS service"
            options={[
                {
                    label: "Amazon S3",
                    value: "s3",
                },
                {
                    label: "AWS Lambda",
                    value: "lambda",
                },
            ]}
        />,
    );
    expect(wrapper.container).toMatchSnapshot();
});
