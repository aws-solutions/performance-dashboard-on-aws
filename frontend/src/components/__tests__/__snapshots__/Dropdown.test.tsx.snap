// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a dropdown with options 1`] = `
<div>
  <div
    class="usa-form-group"
    role="contentinfo"
  >
    <label
      class="usa-label text-bold"
      for="aws-service"
    >
      Select an AWS service
    </label>
    <div
      class="usa-hint"
      id="aws-service-description"
    />
    <select
      aria-describedby="aws-service-description"
      class="usa-select"
      id="aws-service"
      name="aws-service"
    >
      <option
        value="s3"
      >
        Amazon S3
      </option>
      <option
        value="lambda"
      >
        AWS Lambda
      </option>
    </select>
  </div>
</div>
`;

exports[`renders a select with a default value 1`] = `
<div>
  <div
    class="usa-form-group"
    role="contentinfo"
  >
    <label
      class="usa-label text-bold"
      for="aws-service"
    >
      Select an AWS service
    </label>
    <div
      class="usa-hint"
      id="aws-service-description"
    />
    <select
      aria-describedby="aws-service-description"
      class="usa-select"
      id="aws-service"
      name="aws-service"
    >
      <option
        value="s3"
      >
        Amazon S3
      </option>
      <option
        selected=""
        value="lambda"
      >
        AWS Lambda
      </option>
    </select>
  </div>
</div>
`;

exports[`renders a select with error 1`] = `
<div>
  <div
    class="usa-form-group usa-form-group--error"
    role="contentinfo"
  >
    <label
      class="usa-label text-bold"
      for="aws-service"
    >
      Select an AWS service
    </label>
    <div
      class="usa-hint"
      id="aws-service-description"
    />
    <span
      class="usa-error-message"
      id="input-error-message"
      role="alert"
    >
      Something bad happened
    </span>
    <select
      aria-describedby="aws-service-description"
      class="usa-select"
      id="aws-service"
      name="aws-service"
    />
  </div>
</div>
`;

exports[`renders a select with hint 1`] = `
<div>
  <div
    class="usa-form-group"
    role="contentinfo"
  >
    <label
      class="usa-label text-bold"
      for="aws-service"
    >
      Select an AWS service
    </label>
    <div
      class="usa-hint"
      id="aws-service-description"
    >
      Think twice before selecting
    </div>
    <select
      aria-describedby="aws-service-description"
      class="usa-select"
      id="aws-service"
      name="aws-service"
    />
  </div>
</div>
`;

exports[`renders an empty dropdown 1`] = `
<div>
  <div
    class="usa-form-group"
    role="contentinfo"
  >
    <label
      class="usa-label text-bold"
      for="aws-service"
    >
      Select an AWS service
    </label>
    <div
      class="usa-hint"
      id="aws-service-description"
    />
    <select
      aria-describedby="aws-service-description"
      class="usa-select"
      id="aws-service"
      name="aws-service"
    />
  </div>
</div>
`;
